### DISCUSSION TITLE
How do I know if an element is published for the first time?

### DISCUSSION THREAD

**Discussion by boboldehampsink** [craftcms/cms#9638] [Q&A] [answered]
### Description

I used to know this by $event->isNew, but that doesn't work anymore since it already fires at the creation of the first draft.

**Comment by brandonkelly** [comment, marked as answer]
This was an unintended casualty of keeping the entry ID consistent between its unpublished draft and the real entry (#9104). Before 3.7, the unpublished draft was duplicated into a normal entry, with a new ID, therefore `$isNew` was `true`.

It’s too late to revert the change, and I don’t think we should set `$isNew = true` for backwards compatibility since it’s not technically true.

So I just released Craft 3.7.5 with a new `$firstSave` property on entries. It will be `true` whenever an entry is getting saved for the first time as a normal (non-draft, non-revision) entry. (416479f66997bdac211abc883e24d46519a8709e)

**Reply by brandonkelly** [reply]
Just added a section on this to the 3.7 upgrade guide: https://craftcms.com/knowledge-base/upgrading-to-craft-3-7#detecting-newly-saved-entries

**Reply by Anubarak** [reply]
@brandonkelly Maybe you could include in your upgrade guide, that `->isNew` doesn't work as before anymore.
Currently one could think `firstSave` is an additional value / parameter while it in fact it "fixes" a breaking change

**Reply by brandonkelly** [reply]
Fair enough. Just modified the blurb to clarify.

**Comment by BenParizek** [comment]
I've posted some thoughts on [matching New and Updated Element scenarios](https://craftcms.stackexchange.com/a/38169/115) on Stack Exchange which also discusses some considerations like multi-site and resaving elements tasks.

With the number of conditions, it may be nice to see an officially supported `isNewElement` and `isUpdatedElement` method that covers the 80% use case in `craft\helpers\ElementHelper`. I'll propose as a starting point the code I shared in the answer above, and I'd be happy to hear how anybody thinks it could be improved:

```
public static function isNewElement($element, $status = \craft\base\Element::STATUS_ENABLED): bool
{
    return
        $element->firstSave &&
        $element->getIsCanonical() &&
        $element->getStatus() === $status &&
        !ElementHelper::isDraftOrRevision($element) &&
        !$element->resaving &&
        !$element->propagating;
}

public static function isUpdatedElement($element, $status = \craft\base\Element::STATUS_ENABLED): bool
{
    return
        !$element->firstSave &&
        $element->getIsCanonical() &&
        $element->getStatus() === $status &&
        !ElementHelper::isDraftOrRevision($element) &&
        !$element->resaving &&
        !$element->propagating;
}
```

Where we could call it doing something like:

```
\craft\helpers\ElementHelper::isNewElement($element, \craft\elements\Entry::STATUS_LIVE);
```

**Reply by Anubarak** [reply]
Great idea. Thanks for your effort. I would eventually consider to make the status optional because sometimes you want to sync even archived or disabled entries to 3rd party libraries or do certain things independent of the status.

**Reply by brandonkelly** [reply]
For `isNewElement()`…

- `firstSave` will _only_ be `true` for non-draft, non-revision elements, so checking `getIsCanonical()` and `!ElementHelper::isDraftOrRevision()` in addition to `firstSave` are redundant.
  https://github.com/craftcms/cms/blob/972d611d3f38d8ee830df485618d3064e3de4f19/src/services/Elements.php#L2438-L2442
- `firstSave` will _never_ be `true` when resaving elements via a `resave/*` command or `ResaveElements` queue job. Those will only resave existing elements (and won’t alter their state). So checking `!$element->resaving` in addition to `firstSave` is redundant.
- Passing the status into these functions wouldn’t save any complexity – if the calling code has to think about status, it might as well do its own `$element->getStatus() === $status` check.
- If you want to exclude propagating elements, you should be listening to `EVENT_AFTER_PROPAGATE` rather than `EVENT_AFTER_SAVE`, which will only get called once, after the element has been fully saved and propagated.

So that function is completely unnecessary. Just do this:

```php
use craft\base\Element;
use craft\elements\Entry;
use craft\events\ModelEvent;
use yii\base\Event;

Event::on(
    Entry::class,
    Element::EVENT_AFTER_PROPAGATE,
    function(ModelEvent $event) {
        /** @var Entry $entry */
        $entry = $event->sender;
        
        if ($entry->firstSave && $entry->getStatus() === Entry::STATUS_LIVE) {
            // boom
        }
    }
);
```

And for `isUpdatedElement()`…

- All non-drafts/non-revisions are inherently canonical, so calling `$element->getIsCanonical()` in addition to `!ElementHelper::isDraftOrRevision()` is redundant.
- Same points about `getStatus()` and `propagating` as before.

Which leaves us with:

```php
public static function isUpdatedElement($element): bool
{
    return
        !$element->firstSave &&
        !ElementHelper::isDraftOrRevision($element) &&
        !$element->resaving;
}
```

Which is _slightly_ helpful, but only once you’ve taken the time to understand the very specific point of it, which feels like just as much work as adding a couple extra conditions yourself (or creating your own helper method).

**Reply by BenParizek** [reply]
Thanks. It's taken me a while to wrap my head around all the changes and break down my prior assumptions on how it had worked, but it's making more sense now. I've updated my answer on the SE thread to reflect these less redundant checks.
